sharktuner/boo_tuner/boo_tuner.py

```diff
@@ -0,0 +1,333 @@
+    # Split tab-separated arguments (for easier copy-pasting from TSV files).
+    miopen_op_args = [a for arg in miopen_op_args for a in arg.split("\t")]
+
+    # Load MIOpen commands from file if specified, otherwise use command-line arguments.
+    if not commands_file:
+        return [miopen_op_args]
+
+    # Validate that miopen_op_args is empty when using a commands file.
+    if miopen_op_args:
+        raise ValueError(
+            "Cannot specify both --commands-file and MIOpen operation arguments."
+        )
+
+    splitter: Callable[[str], list[str]] = lambda s: (
+        s.strip().split("\t") if commands_file.endswith(".tsv") else shlex.split(s)
+    )
+    with open(commands_file) as f:
+        return [
+            splitter(s) for s in f.readlines() if s.strip() and not s.startswith("#")
+        ]
+
+
+def build_compile_args(compile_command: str, benchmarks_dir: Path) -> list[str]:
+    """Build iree-compile arguments from turbine compile command."""
+    turbine_compile_flags = shlex.split(compile_command)
+
+    # Start with iree-compile and filter out -o flag from turbine flags.
+    compile_args = ["iree-compile"]
+    args_iter = iter(turbine_compile_flags[1:])
+    for arg in args_iter:
+        if arg == "-o":
+            next(args_iter, None)  # Skip the output file path.
+        else:
+            compile_args.append(arg)
+
+    # Add tuner-specific flags.
+    compile_args.extend(
+        [
+            "--iree-config-add-tuner-attributes",
+            "--iree-hal-dump-executable-benchmarks-to",
+            str(benchmarks_dir),
+            "-o",
+            os.devnull,
+        ]
+    )
+
+    return compile_args
+
+
+def main() -> None:
+    parsed_args: tuple[argparse.Namespace, list[str]] = arg_parse()
+    args, miopen_op_args = parsed_args
+    path_config: libtuner.PathConfig = libtuner.PathConfig()
+    path_config.base_dir.mkdir(parents=True, exist_ok=True)
+
+    print("[WARNING] BOO Tuner is still experimental")
+    root_logger = libtuner.setup_logging(args, path_config)
+    print(path_config.run_log, end="\n\n")
+
+    if not args.dry_run:
+        print("Validating devices")
+        libtuner.validate_devices(args.devices)
+        print("Validation successful!\n")
+
+    # These imports are slow due to a pytorch dependency. Keeping them local.
+    # This helps make '--help' fast.
+    from iree.turbine.kernel.boo import runtime as boo_runtime
+    from iree.turbine.kernel.boo.driver.launch import get_launchable
+    from iree.turbine.kernel.boo.op_exports.registry import BooOpRegistry
+
+    logging.getLogger("turbine").setLevel(logging.WARNING)
+
+    assert not (
+        args.commands_file and miopen_op_args
+    ), "Cannot specify both --commands-file and MIOpen operation arguments"
+
+    mio_args = load_commands_from_file_or_args(args.commands_file, miopen_op_args)
+
+    starter_td_spec: Path | None = args.starter_td_spec
+    for idx, cli_args in enumerate(mio_args):
```

Comment:
Can we turn the body of this loop into a helper function? I'm concerned we are putting so much logic directly in `main`.